Compute & forward Ethash proof data

Rizziepit opened the pull request:
This PR finally adds Ethash proof data to the header payload so that Substrate can fully validate and import the header (see Substrate logs at bottom)

There are two main things happening in this PR:
1. I added ethashproof as a git submodule (we've forked it: https://github.com/Snowfork/ethashproof)
2. I added `HeaderChainState` to track... uh.... header state. The biggest piece of work it does is keeping the current + next epoch's cache data in memory for ethashproof

Unfortunately ethashproof uses `fmt.Printf` instead of a logger. So our output is a bit messy and cannot be filtered using log levels:
![Screenshot from 2020-11-23 20-58-24](https://user-images.githubusercontent.com/588701/100050613-a635f900-2dce-11eb-8a43-6a1a515df9ee.png)

Here are the Substrate logs for successful imports:
![Screenshot from 2020-11-23 20-39-49](https://user-images.githubusercontent.com/588701/100049564-7be33c00-2dcc-11eb-8b26-0f7f85748cd3.png)

Rizziepit commented on relayer/chain/ethereum/header_chain_state.go:
```diff
@@ -0,0 +1,143 @@
+// immediately if `number` is in the current or next epoch. Outside that range, it
+// might block for multiple minutes to generate the cache. Calling GetEthashproofCache
+// will also update the current epoch to `number` / 30000.
+func (s *HeaderChainState) GetEthashproofCache(number uint64) (*ethashproof.DatasetMerkleTreeCache, error) {
```
This logic should really have a test. @vgeddes @musnit is there a mocking library you folks use?

philipstanislaus replied:
Agree, tests would be great for this file. 

I prefer https://github.com/stretchr/testify over https://github.com/golang/mock for the better output, but both are fine in my view.

Rizziepit commented on relayer/go.mod:
```diff
@@ -7,7 +7,7 @@ require (
 	github.com/centrifuge/go-substrate-rpc-client v2.0.0-rc6+incompatible
 	github.com/deckarep/golang-set v1.7.1 // indirect
-	github.com/ethereum/go-ethereum v1.9.22
+	github.com/ethereum/go-ethereum v1.8.23
```
This is the highest version that ethashproof builds with :( I didn't want to spend time upgrading ethashproof but happy to create an issue and do this later if you think it's important

philipstanislaus replied:
Since this should be highly security relevant, I feel we need to upgrade ethashproof before going to production. See https://github.com/ethereum/go-ethereum/releases for details on the security issues fixed in previous releases.

musnit replied:
Hm I don't think ethash has changed at all since then but it would still be a good idea to upgrade yeh. we can add a github issue for this to address later.

Rizziepit replied:
https://github.com/Snowfork/polkadot-ethereum/issues/169

philipstanislaus commented on relayer/chain/ethereum/header.go:
```diff
@@ -82,12 +89,45 @@ func MakeHeaderFromEthHeader(gethheader *etypes.Header, log *logrus.Entry) (*cha
 		Seal:             []types.Bytes{mixHashRLP, nonceRLP},
 	}
 
+	// Generate merkle proofs for Ethash
+	indices := ethash.Instance.GetVerificationIndices(
+		blockNumber,
+		ethash.Instance.SealHash(gethheader),
+		gethheader.Nonce.Uint64(),
+	)
+
+	proofData := make([]DoubleNodeWithMerkleProof, len(indices))
+	for i, index := range indices {
+		element, proof, err := ethashproof.CalculateProof(blockNumber, index, proofcache)
+		if err != nil {
+			return nil, err
+		}
+
+		es := element.ToUint256Array()
+		node1Bytes := make([]byte, 64)
+		node2Bytes := make([]byte, 64)
+		// Each 32 byte sequence is left-padded with 0
+		copy(node1Bytes[32-len(es[0].Bytes()):32], es[0].Bytes())
+		copy(node1Bytes[64-len(es[1].Bytes()):], es[1].Bytes())
+		copy(node2Bytes[32-len(es[2].Bytes()):32], es[2].Bytes())
+		copy(node2Bytes[64-len(es[3].Bytes()):], es[3].Bytes())
+		proofH128 := make([][16]byte, len(proof))
+		for j, pr := range proof {
+			proofH128[j] = [16]byte(pr)
+		}
+
+		proofData[i] = DoubleNodeWithMerkleProof{
+			DagNodes: [2]types.H512{
+				types.NewH512(node1Bytes),
+				types.NewH512(node2Bytes),
+			},
+			Proof: proofH128,
+		}
+	}
+
```
Can we put this functionality into a helper function and add a test for it?

Rizziepit replied:
For sure

philipstanislaus commented on relayer/chain/ethereum/header_chain_state.go:
```diff
@@ -0,0 +1,143 @@
+}
+
+// HeaderChainState is a helper for tracking where we are
+// in the Ethereum chain. It warms up caches as we go.
```
As mentioned in #166, how does this logic deal with different forks?

philipstanislaus replied:
I guess the ethash proof logic is fork independent, so that should be fine actually.

philipstanislaus replied:
Would it make sense to separate current/newest block (since that's fork dependent) from the ethashproofCache?

Rizziepit replied:
I think it makes sense to separate tehm. My intuition was to keep this state together since current epoch is related to current forks. But it's not actually relevant to the implementation

philipstanislaus commented on relayer/chain/ethereum/header_chain_state.go:
```diff
@@ -0,0 +1,143 @@
+	// prepareNextEthashproofCache should only ever be called after
+	// nextCache has been set to nil
+	if cacheState.nextCache != nil {
+		return fmt.Errorf("prepareNextEthashproofCache encountered non-nil nextCache")
```
Use logrus here and in other places?

philipstanislaus reviewed (commented):
Great job on the `EthashproofCacheState`, I like how you implemented that 💪 

Left a few comments below.

Rizziepit commented:
The caching logic and proof construction now have decent tests. @philipstanislaus I think I've addressed all your feedback. The only thing I skipped is logging via logrus in `HeaderCacheState`. It felt better to bubble up errors and log them in the listener

philipstanislaus commented:
Cool, will review tomorrow. Thanks!

vgeddes commented:
Is there reason we need to make `ethashproof` a submodule? Can't we just list it as a dependency in `go.mod`, like `github.com/Snowfork/ethashproof`?

vgeddes commented on relayer/chain/ethereum/listener.go:
```diff
@@ -157,7 +172,7 @@ func makeFilterQuery(contracts []Contract) geth.FilterQuery {
 	for _, contract := range contracts {
 		addresses = append(addresses, contract.Address)
-		signature := contract.ABI.Events["AppTransfer"].ID.Hex()
+		signature := contract.ABI.Events["AppTransfer"].Id().Hex()
```
Does the `Id()` method actually exist? I don't see one defined in https://github.com/ethereum/go-ethereum/blob/master/accounts/abi/event.go.

vgeddes commented on .gitmodules:
```diff
@@ -0,0 +1,3 @@
+[submodule "relayer/ethashproof"]
+	path = relayer/ethashproof
+	url = git@github.com:Snowfork/ethashproof.git
```
Is there reason we need to make ethashproof a submodule? Can't we just list it as a dependency in go.mod, like `github.com/Snowfork/ethashproof`? Then everywhere we import `github.com/tranvictor/ethashproof` we would import `github.com/Snowfork/ethashproof` instead?

musnit replied:
ye agree seems better

Rizziepit replied:
No reason we _have_ to make it a submodule. I got the idea from https://github.com/KyberNetwork/bridge_eos_smart_contracts. I think that the monorepo-like situation might benefit us. Two main reasons for this:
* We might want to make changes to ethashproof down the line (definitely for https://github.com/Snowfork/polkadot-ethereum/issues/169), and those changes will have to be in lockstep with the relayer.
* I'm using bits of ethashproof that I would classify as "internals", not the public API. I might reuse bits for transaction proofs too. So managing ethashproof as an internal module is fitting in that sense.

I haven't used submodules before. So I could very well be missing an important consideration. I'm kinda curious to see how this unfolds but will defer to you folks

Rizziepit replied:
Clarification: `github.com/tranvictor/ethashproof` is the module name declared in ethashproof's go.mod file. So I'm assuming I'd need to update ethashproof's go.mod if I want to import it using `github.com/Snowfork/ethashproof`. That's independent of whether ethashproof is a submodule. I wanted to avoid making changes to ethashproof for this PR but I'll update in another PR
